### ISSUE TITLE
[spi_host] Passthrough mode: collisions with software generated commands

### ISSUE THREAD

**Issue by martin-lueker** [lowRISC/opentitan#5134]
In #5077, the specification includes provisions for passthrough mode, where an external hardware block can control SPI_HOST's serial bus.

In this PR spec, this feature currently insists that software must write a bit to surrender control of the serial bus.  Is this the right way to implement this?

- If this software is in control of selecting which hardware controls the bus, is this good enough for our pass-through application?
- If external applications can instantly grab control of the bus, how should we ensure that software commands are not scrambled?
- Are there middle-ground options like a request/grant arbiter?

To me at least the fundamental question comes down to: What are the latency requirements for pass-through mode?

**Comment by eunchan** [comment]
I think it is reasonable to have mode bit as you expected. I think having mode bit then switching to passtrhough and ignore any request from software other than switching the mode to generic mode again is sufficient.

putting arbiter in the spi request will affect the timing significantly (the path is timing restrictive). So simple mux is favorable.

**Comment by tjaychen** [comment]
hey @eunchan , but what about switching from passthrough back to software mode?
we could have a transaction that is ongoing to the serial flash device already right?

**Comment by eunchan** [comment]
> hey @eunchan , but what about switching from passthrough back to software mode?
> we could have a transaction that is ongoing to the serial flash device already right?

I still believe that the spi_host remains in passthrough mode in that scenario. Only SPI_DEVICE internally switch to flash mode to hi-jack the transaction.

Or, if you are talking about the software's mis-behavior, which it switches to sw mode from passthrough by incident, then I think it is inevitable unless the passthrough is controlled by hardware (maybe from SPI_DEVICE?).

**Comment by tjaychen** [comment]
oo sorry i should clarify.

There ARE cases in which software will want to switch away from passthrough to direct control of SPI host.  This for example can happen if a payload were uploaded that software now needs to examine and decide if it should send over to the serial device.

What do you think should happen in this case? It's very possible that upon payload capture, we set busy and block off the external host -> spi device -> spi host path.  But I think you would know best.

**Comment by cfrantz** [comment]
We have this same issue with the interposer on current Google Titan chips.

When using the interposer to connect the SPI_DEVICE through to SPI_HOST within the Root-of-Trust (connecting SPS to SPI using the existing chip's terminology), we must assume that the upstream SPI host (spi master) has complete control over the SPI bus.

Discussing this in terms of the existing Titan chip and how its firmware operates:

There is a control bit in the SPI peripheral which enables or disables passthrough from the SPS peripheral.  I typically refer to this as having the interposer circuit as open (disconnected) or closed (connected).  Once you close the interposer and give control to the upstream device, you must obey SPI conventions or risk causing the upstream device to read corrupted data.

Certain SPI opcodes driven from upstream are expected to execute and return data immediately:
- READ, READ_SFDP, READ_ID, READ_STATUS, EN4B, EX4B, etc.

Some of these opcodes are serviced by dedicated hardware within the SPS peripheral (READ_SFDP, READ_ID, READ_STATUS), while EEPROM READ opcodes are passed through to the SPI peripheral.  EEPROM reads must pass through without delay - there is no "wait" or backpressure mechanism in SPI for READs.

Certain SPI opcodes driven from the upstream device are expected to take time and the upstream device will poll the BUSY bit in the status register (also known as the WIP or work-in-progress bit).  The Titan chip's SPS peripheral has its own local representation of the BUSY to respond to READ_STATUS opcodes.
- ERASE, PROGRAM, etc.

Furthermore, the existing Titan chip has hardware assistance from the SPS peripheral: upon receipt of a certain set of user-specified SPI opcodes (aka "the filter"), the SPS peripheral will set the BUSY bit and interrupt firmware.  Firmware can then open the interposer and drive whatever transactions it wants downstream out of its own SPI host.  Upon completion of this work, firmware can then close the interposer and clear the local representation of the BUSY bit in the SPS peripheral, thus signalling upstream that the long-running opcode is complete and the SPI bus is ready to accept another command.

The existing Titan firmware uses the capabilities of the SPS peripheral to construct a command protocol between the upstream device and the Titan chip itself.  The current firmware can only take ownership of the SPI peripheral in response to a command encoded via one of the BUSY-bit-setting opcodes.

In order for firmware to initiate commands to the downstream SPI EEPROM, the interposer must be open.  When using passthrough, you must use the existing busy-bit conventions of the SPI protocol to not break the upstream device.

**Comment by martin-lueker** [comment]
Resolved.

The existing passthough interface in #5888 and #5936 is sufficient.   Per conversation with @cfrantz and others, There is no use case where FSM driven commands may be interrupted by passthrough commands, as FSM driven commands will only happen when the upstream SPI host (driving the OpenTitan SPI device) is held in reset.

**Comment by tjaychen** [comment]
wait hold on, is that true?

**Comment by martin-lueker** [comment]
Clarification: By SPI Host, I mean not the SPI Host IP, rather the *system level host*.  Is there a  better term for this.

@cfrantz: can you perhaps clarify (here, for posterity) why we are don't care about collisions between simultaneous FSM-driven or passthrough-driven traffic?

**Comment by tjaychen** [comment]
yeah the key paragraph from @cfrantz response to me is here.
Basically, there are going to be situations where certain commands from the external `SPI HOST` will trigger the passthrough path to "open".  This opening then allows the internal spi_host to takeover the pins and directly communicate with the external spi device. 

It might be perfectly to control this timing from the perspective of the `spi_device` and not `spi_host`, but I just wanted to make sure that it's clear. 

```
Furthermore, the existing Titan chip has hardware assistance from the SPS peripheral: upon receipt of a certain set of user-specified SPI opcodes (aka "the filter"), the SPS peripheral will set the BUSY bit and interrupt firmware. Firmware can then open the interposer and drive whatever transactions it wants downstream out of its own SPI host. Upon completion of this work, firmware can then close the interposer and clear the local representation of the BUSY bit in the SPS peripheral, thus signalling upstream that the long-running opcode is complete and the SPI bus is ready to accept another command.
```

**Comment by martin-lueker** [comment]
Great.  However what I'm understanding is for broader system level reasons, OT firmware  will never want to use the internal SPI HOST, at any time when it is concerned about the external SPI Host issuing these "open" commands.    There is no use case where these collisions can occur.  (Or is there?)

**Comment by tjaychen** [comment]
yes there technically shouldn't be any collision.  But there SHOULD be a graceful hand off from..

external host issuing passthrough commands -> spi device "open" passthrough pass (disable it) -> spi host now controls the spi pins

Similarly there probably needs to be a graceful transition in the reverse direction when the software is no longer in need of the interface.

I guess what I'm saying there should not be collision, as long as we provide enough hooks in spi_device / spi_host to inform the software when it is safe for it to take over the pins, and how should it hand control back to passthrough when it is done. 

Does that make sense? Sorry if I'm confusing things.

**Comment by martin-lueker** [comment]
Sure, but isn't all of this outside the scope of the SPI HOST IP?

**Comment by tjaychen** [comment]
i.. THINK so :)
but it depends on where you want to put the logic i guess.

You could argue for example, the `spi_host` ip should be the final arbitrator, and that the passthrough_en coming from `spi_device` is just a request signal.  The "close" switch then, could then be implemented on the spi_host side (you could also allocate a software bit to say "i no longer need this interface", similar to its request signal). 

This could also be done on the spi_device side as well, but I just really wanted to make sure that the overall scheme is clear.

**Comment by martin-lueker** [comment]
FYI @eunchan

SPI_DEVICE is the only and final arbiter.  SPI_HOST IP really has no knowledge of what is happening between SPI_DEVICE and the upstream host.   From what I understand there is really nothing that SPI_HOST can or should do to release this (and this would be an unnecessary complication).  Furthermore if a condition arises where SPI DEVICE needs to grab the bus immediately while SPI_HOST IP is transacting, then there is no remedy.

All of this needs to be managed in Firmware.

**Comment by tjaychen** [comment]
that all sounds good to me. 
So maybe the right thing to do here is really open a separate `SPI_DEVICE` issue to track this?

**Comment by cfrantz** [comment]
This last comment by @martin-lueker is correct.

My reading of the firmware which implements the interposer in the legacy chip is that we rely on the behavior of the upstream (external) spi host driving a transaction into the spi_device to behave properly.  When an upstream device issues an opcode which is trapped by firmware (e.g. ERASE, PROGRAM), the spi eeprom protocol around the busy bit is what implicitly gives firmware the capability to use the internal spi_host to drive an appropriate transaction to the backend eeprom.

My recollection of the existing chip (and the code supports this) is that when the spi_device captures a command and allows firmware to handle it, the interposer is considered "open" (or disconnected).  The spi_host driver checks for this condition before issuing a command to the hardware to prevent an accidental collision (the check is `if (passthrough_enabled && !busy_bit) return ERROR_BUSY;`).

**Comment by tjaychen** [comment]
sounds good! I think specifically to your last point @cfrantz , that's something we'll have to make sure to test end-to-end in all of our platforms, so that we can ensure when the host is well behaved, everything works correctly together.

**Comment by tjaychen** [comment]
i'll also open a separate issue on spi_device to point out we need to do this kind of arbitration, and reference the comments in this one.

**Comment by tjaychen** [comment]
@eunchan i think we can close this now right? This is something that's being handled on spi_device side.
I know we had some discussions about it, but i can't completely remember the resolution.

**Comment by martin-lueker** [comment]
Hi all (including @eunchan), just another ping here.  Can we close this one?

**Comment by tjaychen** [comment]
@eunchan do you mind refreshing our memories :)

**Comment by tjaychen** [comment]
@eunchan i took a quick look, and i think `passthrough_en` right now is based on the mode only, which is programmed by software. 

I think this means that if a command / payload is captured by software, it will be firmware responsibility to flip the bit. This means it is possible to flip it in the middle of a transaction.

By my reading this is "probably" okay, since the `spi_host` will just directly take control of the mux, and this SHOULD mean `csb` deasserts. So as long as our passthrough filtering was working correctly to start with, this shouldn't be an issue.

@a-will any thoughts?

**Comment by martin-lueker** [comment]
To reiterate the points above, there is no back pressure in SPI and when SPI passthrough is needed there is _no time_ to coordinate a shutdown of the transaction.   This really can only be resolved by a SPI_DEVICE FW programming model which realizes that there is no way these two things can be allowed to interact.   A register switch could be added to lock things out to stop glitches from bad firmware, but really that is the best one can do. If buggy firmware exists then there will still be lost transactions.

Given the mountains of work to be done in <3 weeks, I would like to icebox this issue.  This is now a feature request.

**Comment by a-will** [comment]
The case of being able to re-enable passthrough in the middle of an incoming command is not a firmware fault. It would be a bug in the hardware. Firmware can't do anything to ensure the switch happens on command boundaries.

If reads are suddenly enabled in the middle of a command, the external SPI flash may see a different opcode, and actual stateful changes to the external SPI flash may occur (though probably limited in scope).

**Comment by tjaychen** [comment]
> It seems a little janky to not have `passthrough_en` shut off when the `busy` bit is set. Does that create some potential for glitches to the external SPI flash?
> 
> Similarly, we'd probably not want to start doing passthrough in the middle of a command. It looks like if we have to change the mode, `passthrough_en` can go high _at any time_, sending through bits starting, say, in the middle of the opcode. That may be a bug--Perhaps we need a more explicit handshake with software?

my preference is to have a more sequenced switch as well. Especially since, as @cfrantz pointed out above, this usually happens during a busy poll, we can align to transaction ends to make changes (so make sure we only change when things are idle).

**Comment by a-will** [comment]
> @martin-lueker to clarify one thing. This isn't a "spi-host" issue. It's purely on how spi device controls the passthrough_en timing based on where the command boundaries are. Right now it could be anywhere, I think what @a-will and I are saying it's better that it's defined.

Sorry we kind of corrupted your original issue ;) Changes would all need to go in `spi_device`.

`spi_host` is good with just the `passthrough_en` signal. If you wanted to isolate `spi_host` from bugs in `spi_device`, then a single software register that gates the `passthrough_en` signal would do the trick. But I don't think that's necessary.

**Comment by martin-lueker** [comment]
> @martin-lueker to clarify one thing. This isn't a "spi-host" issue. It's purely on how spi device controls the passthrough_en timing based on where the command boundaries are. Right now it could be anywhere, I think what @a-will and I are saying it's better that it's defined.

I see, well can we rename the issue then?

**Comment by tjaychen** [comment]
> > The case of being able to re-enable passthrough in the middle of an incoming command is not a firmware fault. It would be a bug in the hardware. Firmware can't do anything to ensure the switch happens on command boundaries.
> > If reads are suddenly enabled in the middle of a command, the external SPI flash may see a different opcode, and actual stateful changes to the external SPI flash may occur (though probably limited in scope).
> 
> I agree that bad things could happen.
> 
> Well, the entire design is predicated on the assumption that these modes are not needed back to back (see above). Passthrough is needed by other hosts on the system, and when FSM driven transactions are needed, the other hosts are in reset or not present. Under that use case, firmware would be in error to request FSM driven operation any time that passthrough could be invoked.
> 
> If there are new use-cases could we please document them?

unfortunately this is definitely NOT the case. (i believe @cfrantz explained above).  Passthrough operation and normal FSM driven operation are intended to co-exist.  As an example when the host issues an erase operation, it will, when behaving correctly, go into a busy poll loop.  The device is then expected to switch to FSM driven operation to handle the transaction that was just received.  During this entire time, the host will continue polling the device, it is neither in reset nor absent. So we do have to handle these transition cases somehow.

Either we convince ourselves it is safe for software to do so, or we need enhance the hw to be able to handle it.

**Comment by eunchan** [comment]
My expectation was (I am not sure I understand the issue correctly), when SW wants the `busy` is set, SW may configure filtering on.

But the issue @a-will ask is, the following commands right? (probably Read Status?). The status checking falls into the SPI_DEVICE. So expected to be filtered to the downstream device. But you are right. There're collision cases.

So, looks like it is safe to turn off passthrough_en by `is_active && (busy_ctrl_en ? !busy : 1'b 1)` . We can add `busy_ctrl_en` in CONTROl CSR. WDYT @tjaychen @a-will

**Comment by a-will** [comment]
> So, looks like it is safe to turn off passthrough_en by `is_active && (busy_ctrl_en ? !busy : 1'b 1)` . We can add `busy_ctrl_en` in CONTROl CSR. WDYT @tjaychen @a-will

Hm... I suspect it won't be entirely that straightforward. It's a bit of a struggle to get everything latched into the SCK domain, due to the dearth of clock edges. @tjaychen has good ideas for handling that transfer, though.

**Comment by eunchan** [comment]
> Hm... I suspect it won't be entirely that straightforward. It's a bit of a struggle to get everything latched into the SCK domain, due to the dearth of clock edges. @tjaychen has good ideas for handling that transfer, though.
>
busy signal in the current SPI_DEVICE is maintained in peripheral clock domain and also SCK domain. First, when SPI_DEVICE receives a command setting busy, the IP sets SCK domain busy signal first. It is propagated into other modules in SPI_DEVICE. Then, when the transaction ends, the signal is then latched into the peripheral clock domain, which is visible to the SW. When SW clears the BUSY signal, the cleared signal is latched to SCK at the next transaction.

So, the assumption of host reading the STATUS, it should be OK. But if somehow host system sends the admissable command to the downstream flash device right away after BUSY is cleared, it will fail as latching to SCK takes two ~ three cycles. (during that time, `passthrough_en` is turned off)

**Comment by a-will** [comment]
> busy signal in the current SPI_DEVICE is maintained in peripheral clock domain and also SCK domain. First, when SPI_DEVICE receives a command setting busy, the IP sets SCK domain busy signal first. It is propagated into other modules in SPI_DEVICE. Then, when the transaction ends, the signal is then latched into the peripheral clock domain, which is visible to the SW. When SW clears the BUSY signal, the cleared signal is latched to SCK at the next transaction.
> 
> So, the assumption of host reading the STATUS, it should be OK. But if somehow host system sends the admissable command to the downstream flash device right away after BUSY is cleared, it will fail as latching to SCK takes two ~ three cycles. (during that time, `passthrough_en` is turned off)

Oh, great! You've already thought all of that through. I think we'll be okay if the host suddenly performs some other command -- That's either a violation of protocol, or the host rebooted or something. For the latter case, we'll almost certainly see a non-pass-through command first, like Read (JEDEC) ID or Read SFDP, something to query our device.

**Comment by tjaychen** [comment]
> My expectation was (I am not sure I understand the issue correctly), when SW wants the `busy` is set, SW may configure filtering on.
> 
> But the issue @a-will ask is, the following commands right? (probably Read Status?). The status checking falls into the SPI_DEVICE. So expected to be filtered to the downstream device. But you are right. There're collision cases.
> 
> So, looks like it is safe to turn off passthrough_en by `is_active && (busy_ctrl_en ? !busy : 1'b 1)` . We can add `busy_ctrl_en` in CONTROl CSR. WDYT @tjaychen @a-will

@eunchan would this `busy` be the one maintained in peripheral clock domain..?

**Comment by eunchan** [comment]
> 



> > My expectation was (I am not sure I understand the issue correctly), when SW wants the `busy` is set, SW may configure filtering on.
> > But the issue @a-will ask is, the following commands right? (probably Read Status?). The status checking falls into the SPI_DEVICE. So expected to be filtered to the downstream device. But you are right. There're collision cases.
> > So, looks like it is safe to turn off passthrough_en by `is_active && (busy_ctrl_en ? !busy : 1'b 1)` . We can add `busy_ctrl_en` in CONTROl CSR. WDYT @tjaychen @a-will
> 
> @eunchan would this `busy` be the one maintained in peripheral clock domain..?

Hum... That would be good for blocking the passthrough interface (assume no CDC protection  here). Releasing (clearing the BUSY) then would be the SW responsibility to not clear BUSY when the host system tries to send SPI transactions.

I will revise the SPI_DEVICE to support this.

**Comment by a-will** [comment]
> > 
> 
> > > My expectation was (I am not sure I understand the issue correctly), when SW wants the `busy` is set, SW may configure filtering on.
> > > But the issue @a-will ask is, the following commands right? (probably Read Status?). The status checking falls into the SPI_DEVICE. So expected to be filtered to the downstream device. But you are right. There're collision cases.
> > > So, looks like it is safe to turn off passthrough_en by `is_active && (busy_ctrl_en ? !busy : 1'b 1)` . We can add `busy_ctrl_en` in CONTROl CSR. WDYT @tjaychen @a-will
> > 
> > 
> > @eunchan would this `busy` be the one maintained in peripheral clock domain..?
> 
> Hum... That would be good for blocking the passthrough interface (assume no CDC protection here). Releasing (clearing the BUSY) then would be the SW responsibility to not clear BUSY when the host system tries to send SPI transactions.

That is impossible. The host can send SPI transactions whenever it wants. Changes to passthrough_en should be on command boundaries only, done in the SPI clock domain, else we get the potential for corruption.

**Comment by tjaychen** [comment]
i think we can officially claim this issue is addressed with #11456 !
let me know if you disagree @eunchan / @a-will , i will close for now.
